perception/map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -0,0 +1,997 @@
+}
+
+std::vector<PredictedRefPath> MapBasedPredictionNode::getPredictedReferencePath(
+  const TrackedObject & object, const LaneletsData & current_lanelets_data,
+  const double object_detected_time)
+{
+  const double delta_horizon = 1.0;
+  const double obj_vel = object.kinematics.twist_with_covariance.twist.linear.x;
+
+  std::vector<PredictedRefPath> all_ref_paths;
+  for (const auto & current_lanelet_data : current_lanelets_data) {
+    // Step1. Get the path
+    // Step1.1 Get the left lanelet
+    lanelet::routing::LaneletPaths left_paths;
+    auto opt_left = routing_graph_ptr_->left(current_lanelet_data.lanelet);
+    if (!!opt_left) {
+      for (double horizon = prediction_time_horizon_; horizon > 0; horizon -= delta_horizon) {
+        const double search_dist = horizon * obj_vel + 10.0;
+        lanelet::routing::LaneletPaths tmp_paths =
+          routing_graph_ptr_->possiblePaths(*opt_left, search_dist, 0, false);
+        addValidPath(tmp_paths, left_paths);
+      }
+    }
+
+    // Step1.2 Get the right lanelet
+    lanelet::routing::LaneletPaths right_paths;
+    auto opt_right = routing_graph_ptr_->right(current_lanelet_data.lanelet);
+    if (!!opt_right) {
+      for (double horizon = prediction_time_horizon_; horizon > 0; horizon -= delta_horizon) {
+        const double search_dist = horizon * obj_vel + 10.0;
+        lanelet::routing::LaneletPaths tmp_paths =
+          routing_graph_ptr_->possiblePaths(*opt_right, search_dist, 0, false);
+        addValidPath(tmp_paths, right_paths);
+      }
+    }
+
+    // Step1.3 Get the centerline
+    lanelet::routing::LaneletPaths center_paths;
+    for (double horizon = prediction_time_horizon_; horizon > 0; horizon -= delta_horizon) {
+      const double search_dist = horizon * obj_vel + 10.0;
+      lanelet::routing::LaneletPaths tmp_paths =
+        routing_graph_ptr_->possiblePaths(current_lanelet_data.lanelet, search_dist, 0, false);
+      addValidPath(tmp_paths, center_paths);
+    }
+
+    // Skip calculations if all paths are empty
+    if (left_paths.empty() && right_paths.empty() && center_paths.empty()) {
+      continue;
+    }
+
+    // Step2. Predict Object Maneuver
+    const Maneuver predicted_maneuver =
+      predictObjectManeuver(object, current_lanelet_data, object_detected_time);
+
+    // Step3. Allocate probability for each predicted maneuver
+    const auto maneuver_prob =
+      calculateManeuverProbability(predicted_maneuver, left_paths, right_paths, center_paths);
+
+    // Step4. add candidate reference paths to the all_ref_paths
+    const float path_prob = current_lanelet_data.probability;
+    addReferencePaths(
+      object, left_paths, path_prob, maneuver_prob, Maneuver::LEFT_LANE_CHANGE, all_ref_paths);
+    addReferencePaths(
+      object, right_paths, path_prob, maneuver_prob, Maneuver::RIGHT_LANE_CHANGE, all_ref_paths);
+    addReferencePaths(
+      object, center_paths, path_prob, maneuver_prob, Maneuver::LANE_FOLLOW, all_ref_paths);
```

Comment:
FYI. When you call a function several times with almost the same args, make a wrapper could improve the readability. This code would be like below.

```
    const auto addReferencePathsLocal = [&](const auto & paths, const auto & maneuver) {
      addReferencePaths(object, paths, path_prob, maneuver_prob, maneuver, all_ref_paths);
    };
    addReferencePathsLocal(left_paths, Maneuver::LEFT_LANE_CHANGE);
    addReferencePathsLocal(right_paths, Maneuver::RIGHT_LANE_CHANGE);
    addReferencePathsLocal(center_paths, Maneuver::LANE_FOLLOW);
```

Reply:
done